windows-gaming-driver/src/setup/usb.rs

```diff
@@ -103,51 +203,42 @@ fn pick(special: Option<HidKind>,
     let mut devs = Vec::new();
 
     for dev in iter.scan_devices()? {
-        let dev = if special.is_some() {
+        let dev = if kind.is_some() {
             dev.parent().unwrap()
         } else {
             dev
         };
 
-        let mut vendor = None;
-        let mut product = None;
-        let mut vendor_name = "Unknown vendor".to_owned();
-        let mut product_name = "Unknown product".to_owned();
-
-        for attr in dev.properties() {
-            if let Some(val) = attr.value().to_str() {
-                if attr.name() == "ID_VENDOR_ID" {
-                    vendor = parse_hex(val);
-                } else if attr.name() == "ID_MODEL_ID" {
-                    product = parse_hex(val);
-                } else if attr.name() == "ID_VENDOR_FROM_DATABASE" {
-                    vendor_name = val.to_owned();
-                } else if attr.name() == "ID_MODEL_FROM_DATABASE" {
-                    product_name = val.to_owned();
-                }
-            }
-        }
+        let vendor = dev.property_value("ID_VENDOR_ID").map(|s| s.to_str().unwrap()).and_then(parse_hex);
+        let product = dev.property_value("ID_MODEL_ID").map(|s| s.to_str().unwrap()).and_then(parse_hex);
+        let vendor_name = dev.property_value("ID_VENDOR_FROM_DATABASE").map(|s| s.to_string_lossy().into_owned());
+        let product_name = dev.property_value("ID_MODEL_FROM_DATABASE").map(|s| s.to_string_lossy().into_owned());
+        let busnum = dev.attribute_value("busnum").map(|s| s.to_str().unwrap()).and_then(parse_hex);
+        let devpath = dev.attribute_value("devpath").map(|s| s.to_str().unwrap()).and_then(parse_hex);
 
         let id = (vendor.unwrap(), product.unwrap()).into();
-        if !blacklist.contains(&UsbBinding::ById(id)) {
-            println!("[{}]\t{} {} [{:04x}:{:04x}]", devs.len(), vendor_name, product_name,
-                     id.vendor, id.product);
-            devs.push(Some(id));
-        }
-    }
-
-    if allow_abort {
-        println!("[{}]\tNone of the above", devs.len());
-        devs.push(None);
+        let port = (busnum.unwrap(), devpath.unwrap()).into();
+        let kind = if dev.devtype() == Some(OsStr::new("usb_interface"))
```

Comment:
This part is broken. Perhaps you refactored it without thinking about what it does?

I assume what you're trying to do here is classify devices into mouse/keyboard/unknown for full device listings. Otherwise you already know that you're only going to get mouses or keyboards so this code is a bit pointless. In that case though, this piece of code will *always* classify as unrelated. Why?

Well, the reason why we filter `DEVTYPE` for either `usb_device` or `usb_interface` above is that `usb_device` is what we're looking for (a USB device with vendor and model id) but filtering for HIDs is only possible at the `usb_interface` level (since one device can be both a mouse and a keyboard, for example). That's also why we need to select the **parent** device in that case: We're filtering for interfaces but what we need is devices.

So, to make this piece of code useful, you would have to return a *list* of `HidKind`s (since each device can have multiple interfaces), then query for the device's *children* and examine those.